src/abstract_charm.py

```diff
@@ -324,15 +352,29 @@ def reconcile(self, event=None) -> None:  # noqa: C901
                     and isinstance(workload_, workload.AuthenticatedWorkload)
                     and workload_.container_ready
                 ):
-                    self._reconcile_node_port(event=event)
+                    self._reconcile_service()
                     self._database_provides.reconcile_users(
                         event=event,
-                        router_read_write_endpoint=self._read_write_endpoint,
-                        router_read_only_endpoint=self._read_only_endpoint,
+                        router_read_write_endpoints=self._read_write_endpoints,
+                        router_read_only_endpoints=self._read_only_endpoints,
                         exposed_read_write_endpoint=self._exposed_read_write_endpoint,
                         exposed_read_only_endpoint=self._exposed_read_only_endpoint,
                         shell=workload_.shell,
                     )
+                    # _ha_cluster only assigned a value in machine charms
+                    if self._ha_cluster:
+                        self._database_provides.update_endpoints(
+                            router_read_write_endpoints=self._read_write_endpoints,
+                            router_read_only_endpoints=self._read_only_endpoints,
+                            exposed_read_write_endpoint=self._exposed_read_write_endpoint,
+                            exposed_read_only_endpoint=self._exposed_read_only_endpoint,
+                        )
+                    elif self._check_service_connectivity():
```

Comment:
opted to use an abstractmethod inhttps://github.com/canonical/mysql-router-k8s-operator/pull/328/commits/c6f686906759d0aa5072864b9c40fb1273d4aaa0 -> this will help abstract out different logic for vm and k8s